javascript/app-event-analysis/src/components/LineChart/hooks/useMouseHandlers.ts

```diff
@@ -0,0 +1,330 @@
+/*!
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project.
+ */
+import { select } from 'd3'
+import { useCallback, useMemo } from 'react'
+
+import { useTooltip } from '../../../hooks/useTooltip.js'
+import { useTreatedUnitsMap } from '../../../hooks/useTreatedUnitsMap.js'
+import type { LineData, TooltipInfo } from '../../../types'
+import {
+	HIGHLIGHT_LINE,
+	LINE_ELEMENT_CLASS_NAME,
+	LINE_WIDTH,
+	LINE_WIDTH_HOVER,
+	LINE_WIDTH_TREATED,
+	OUTPUT_LINE,
+	OUTPUT_LINE_WIDTH,
+	SYNTHETIC_UNIT,
+	TRANSPARENT_LINE,
+} from '../../../types'
+import { getHoverIdFromValue } from '../../../utils/charts.js'
+import { isValidUnit } from '../../../utils/validation.js'
+import type {
+	HandleLineMouseClickOrMoveProps,
+	MouseHandlers,
+	MouseHandlersProps,
+} from '../LineChart.types.js'
+import { constructLineTooltipContent } from '../LineChart.utils.js'
+
+export function useMouseHandlers(props: MouseHandlersProps): MouseHandlers {
+	const tooltip = useTooltip()
+	const { hide: hideTooltip } = tooltip
+
+	const fullProps = useMemo<HandleLineMouseClickOrMoveProps>(
+		() => ({ ...props, tooltip }),
+		[tooltip, props],
+	)
+
+	const handleContainerClick = useHandleContainerClick(fullProps)
+	const handleLineMouseClick = useHandleLineMouseClick(fullProps)
+	const handleLineMouseMove = useHandleLineMouseMove(fullProps)
+	const handleLineMouseLeave = useHandleLineMouseLeave(fullProps)
+
+	return useMemo(() => {
+		return {
+			tooltip,
+			handleLineMouseMove,
+			handleLineMouseClick,
+			handleLineMouseLeave,
+			handleContainerClick,
+			handleClickOutside: () => hideTooltip(true),
+		}
+	}, [
+		tooltip,
+		handleLineMouseMove,
+		handleLineMouseClick,
+		handleLineMouseLeave,
+		handleContainerClick,
+		hideTooltip,
+	])
+}
+
+function useHandleContainerClick(props: HandleLineMouseClickOrMoveProps) {
+	const { setHoverUnit, tooltip } = props
+	const { hide: hideTooltip } = tooltip
+	return useCallback(
+		(event: React.MouseEvent) => {
+			const element = event.target as Element
+
+			if (!element.classList.contains(LINE_ELEMENT_CLASS_NAME)) {
+				hideTooltip(true)
+				setHoverUnit('')
+			}
+		},
+		[hideTooltip, setHoverUnit],
+	)
+}
+
+function useHandleLineMouseLeave(props: HandleLineMouseClickOrMoveProps) {
```

Comment:
It's not worth refactoring this at the moment, but ultimately it would be much cleaner if the event handlers just set hover or selection state, and then a re-render loop captures the properties automatically instead of having the callback actually manipulate the content.